### PR TITLE
[PyTorch] faster empty_tensor_restride via smarter refresh_contiguous

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49371 [PyTorch] Use .sizes() isntead of .size() in cat_serial_kernel_impl
* #49368 [PyTorch] Use .sizes() instead of .size() in _cat_out_cpu
* #49364 [PyTorch] Avoid extra Tensor refcounting in _cat_out_cpu
* #49359 [PyTorch] Merge CoinflipTLS into RecordFunctionTLS
* #49355 [PyTorch] Avoid move-constructing a List in listConstruct
* **#49349 [PyTorch] faster empty_tensor_restride via smarter refresh_contiguous**
* #49264 [PyTorch] Make tls_local_dispatch_key_set inlineable
* #49263 [PyTorch] Make TORCH_CHECK less likely to interfere with inlining
* #49262 [PyTorch] native_cpp_binding for size() and stride()

`empty_tensor_restride` knows what memory format it just
restrided (restrode?) to. Don't throw that information away.

Differential Revision: [D25539840](https://our.internmc.facebook.com/intern/diff/D25539840/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25539840/)!

### CODE REVIEW

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -1573,6 +1574,42 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
...
+   * we know the current MemoryFormat.
+   */
+  void refresh_contiguous(MemoryFormat memory_format) {
+    // NOTE: Make sure to keep the other overload in sync with this implementation!
+    is_contiguous_ = memory_format == MemoryFormat::Contiguous || compute_contiguous();
+    switch (memory_format) {
+      case MemoryFormat::Contiguous:
+        is_channels_last_contiguous_ = false;
+        is_channels_last_contiguous_ = false;
+        is_channels_last_3d_contiguous_ = false;
```
Comment (ngimel) [inline c10/core/TensorImpl.h:1588]: `is_coniguous_` and `is_channels_last_contiguous_` are not mutually exclusive, e.g. contiguous tensor of size (N,C,1,1) has both as `true`.

Comment (swolchok) [reply c10/core/TensorImpl.h:1588]: is there anything useful about this diff, then? I don't understand all the subtleties here.

**Comment by swolchok** [comment]
abandoning, not as obvious as I thought

File: c10/core/TensorImpl.h (hunk 1 above)
Comment (ngimel) [reply c10/core/TensorImpl.h:1588]: There are useful things, contiguous computations are repeated twice now, and they shouldn't be. But you are right, it's more annoying. I should still do it.

**Pull request closed without merging**